pdfplumber/page.py

```diff
@@ -115,6 +119,42 @@ def normalize_color(
     return separate_pattern(tuplefied)
 
 
+class PDFPageAggregatorWithMarkedContent(PDFPageAggregator):
+    """Extract layout from a specific page, adding marked-content IDs to
+    objects where found."""
+
+    cur_mcid: Optional[int] = None
+
+    def begin_tag(self, tag: PSLiteral, props: Optional[PDFStackT] = None) -> None:
+        if isinstance(props, dict) and "MCID" in props:
+            self.cur_mcid = props["MCID"]
+        else:
+            self.cur_mcid = None
+
+    def end_tag(self) -> None:
+        self.cur_mcid = None
+
+    def tag_cur_item(self, item_type: Any) -> None:
+        # Implementation Inheritance Considered Harmful
+        cur_obj = self.cur_item._objs[-1]
+        assert isinstance(cur_obj, item_type)
```

Comment:
I'm curious: What is the bigger-picture purpose of this assert statement?

Reply:
Ah, it (and the associated `item_type` parameter) is just there to confirm my understanding of how `pdfminer.six` works internally - as the comment above notes this depends on the internal behaviour of the `render_*` methods.  In an ideal world they would simply return the objects they create instead of modifying some internal state.